ctlcmd/cmdbase.py

```diff
@@ -771,7 +772,198 @@ def globcomp(text):
+  def __init__(self, cmd):
+    print("initialize rootfilecmd")
+    controlcmd.__init__(self, cmd)
+    
+  def add_args(self):
+    print("add_args in rootfilecmd")
+    group = self.parser.add_argument_group(
+        "root file saving options", """Options for changing the root file location.
+        For more details, see the official documentation.""")
+    group.add_argument('--saveroot',
+                       type=str,
+                       default=self.DEFAULT_ROOTFILE,
+                       help="""File path to save (placeholders in angle braces).
+                       default=%(default)s""")
+ 
+  def parse(self,args):
+    print("parse in rootfilecmd")
+    if not args.saveroot:  # Early exit if savefile is not set
+      self.saveroot = None
+      return args
+    filename = args.saveroot
+
+    # Adding time stamp filenames
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    filename = re.sub('<TIMESTAMP>', timestring, filename, flags=re.IGNORECASE)
+
+    # Adding boardid to the settings
+    filename = re.sub('<BOARDID>',
+                      str(self.board.boardid),
+                      filename,
+                      flags=re.IGNORECASE)
+    # Adding boardtype to the settings
+    filename = re.sub('<BOARDTYPE>',
+                      str(self.board.boardtype),
+                      filename,
+                      flags=re.IGNORECASE)
 
+    # Substituting tokens for argument values
+    for action in self.parser._actions:
+      for string in action.option_strings:
+        string = string.strip('-')
+        if not hasattr(args, string): continue
+        val = ''
+        try:
+          val = action.type(getattr(args, string))
+        except:
+          pass
+        substring = '{0}{1}'.format(string, val)
+        substring = re.sub(r'\.', 'p', substring)
+        filename = re.sub('\\<{0}\\>'.format(string),
+                          substring,
+                          filename,
+                          flags=re.IGNORECASE)
+      self.saveroot=filename
+      return args
+      
+  def openroot(self,functiontype):
+    print("openroot rootfilecmd")
+    file = uproot.recreate(self.saveroot)
+    self.functiontype=functiontype
+    
+    halign_titles=["humival","uncval"]
+    lowlight_collect_titles=["readout"]
+    timescan_titles=["lumival","uncval","S2","S4"]
+    tb_levelped_titles=[""]
+    visualhscan_titles=["center x","center y"]
+    visualzscan_titles=["laplace","center x","center y","center area","center maxmeas"]
+    self.standardtitles=["time","det_id","gantry x","gantry y","gantry z","LED bias voltage","LED temp","SiPM temp"]
+    self.titles = [halign_titles,lowlight_collect_titles,timescan_titles,tb_levelped_titles,visualhscan_titles,visualzscan_titles]
+    
+    if functiontype == "halign"or"zscan":
+      file.mktree("DataTree", {self.standardtitles[0]:np.float32,self.standardtitles[1]:np.int_,self.standardtitles[2]:np.float32,
```

Comment:
Best practice is not to have inheritance specific results defined in the base class. What we would want is to define a new `rootfilecmd.treestruture()` method that returns the dictionary required for the standard tree structure, and have subsequent command classes overload this method.

Reply:
Alternatively, we can leave the tree structure blank here, and create the tree structure on the first call of the `fill` method. (You can check if a file is empty using the `file.keys()` method.)